diff --git a/scripts/xdg-open.in b/scripts/xdg-open.in
index 0fbd63d..4baa152 100644
--- a/scripts/xdg-open.in
+++ b/scripts/xdg-open.in
@@ -286,7 +286,7 @@ search_desktop_file()
 open_generic_xdg_mime()
 {
     filetype="$2"
-    default=`xdg-mime query default "$filetype"`
+    default=`xdg-mime query default "$filetype" | sed "s/;.*//"`
     if [ -n "$default" ] ; then
         xdg_user_dir="$XDG_DATA_HOME"
         [ -n "$xdg_user_dir" ] || xdg_user_dir="$HOME/.local/share"
