--- a/src/plugins/platforms/xcb/qxcbkeyboard.cpp.orig	2016-12-01 17:17:04.000000000 +0900
+++ b/src/plugins/platforms/xcb/qxcbkeyboard.cpp	2016-12-20 14:57:16.300329932 +0900
@@ -980,12 +980,14 @@
         return QList<int>();
     }
     // get kb state from the master xkb_state and update the temporary kb_state
+    xkb_layout_index_t depressedLayout = xkb_state_serialize_layout(xkb_state, XKB_STATE_LAYOUT_DEPRESSED);
+    xkb_layout_index_t latchedLayout = xkb_state_serialize_layout(xkb_state, XKB_STATE_LAYOUT_LATCHED);
     xkb_layout_index_t lockedLayout = xkb_state_serialize_layout(xkb_state, XKB_STATE_LAYOUT_LOCKED);
     xkb_mod_mask_t latchedMods = xkb_state_serialize_mods(xkb_state, XKB_STATE_MODS_LATCHED);
     xkb_mod_mask_t lockedMods = xkb_state_serialize_mods(xkb_state, XKB_STATE_MODS_LOCKED);
     xkb_mod_mask_t depressedMods = xkb_state_serialize_mods(xkb_state, XKB_STATE_MODS_DEPRESSED);
 
-    xkb_state_update_mask(kb_state, depressedMods, latchedMods, lockedMods, 0, 0, lockedLayout);
+    xkb_state_update_mask(kb_state, 0, latchedMods, lockedMods, depressedLayout, latchedLayout, lockedLayout);
     quint32 keycode = event->nativeScanCode();
     // handle shortcuts for level three and above
     xkb_layout_index_t layoutIndex = xkb_state_key_get_layout(kb_state, keycode);
@@ -1039,7 +1041,7 @@
                     depressed |= (1 << controlMod);
                 if (metaMod < 32 && neededMods & Qt::MetaModifier)
                     depressed |= (1 << metaMod);
-                xkb_state_update_mask(kb_state, depressed, latchedMods, lockedMods, 0, 0, lockedLayout);
+                xkb_state_update_mask(kb_state, depressed, latchedMods, lockedMods, depressedLayout, latchedLayout, lockedLayout);
                 sym = xkb_state_key_get_one_sym(kb_state, keycode);
             }
             if (sym == XKB_KEY_NoSymbol)
@@ -1507,12 +1509,7 @@
     // used for standard shortcuts in checks like "key == QKeySequence::Copy",
     // users can still see the actual X11 keysym with QKeyEvent::nativeVirtualKey
     Qt::KeyboardModifiers modifiers = translateModifiers(state);
-    xcb_keysym_t translatedSym = XKB_KEY_NoSymbol;
-    if (modifiers & Qt::ControlModifier && !isLatin(sym))
-        translatedSym = lookupLatinKeysym(code);
-    if (translatedSym == XKB_KEY_NoSymbol)
-        translatedSym = sym;
-    int qtcode = keysymToQtKey(translatedSym, modifiers, string);
+    int qtcode = keysymToQtKey(sym, modifiers, string);
 
     bool isAutoRepeat = false;
     if (type == QEvent::KeyPress) {
