diff --git a/png.c b/png.c
index 88cf9e7..e909dd3 100644
--- a/png.c
+++ b/png.c
@@ -2394,8 +2394,10 @@ png_compare_ICC_profile_with_sRGB(png_const_structrp png_ptr,
                       * discourage their use, skip the 'have_md5' warning below,
                       * which is made irrelevant by this error.
                       */
+#if 0
                      png_chunk_report(png_ptr, "known incorrect sRGB profile",
                          PNG_CHUNK_ERROR);
+#endif
                   }
 
                   /* Warn that this being done; this isn't even an error since
