diff --git a/platforms/unix/platform_unix.cpp b/platforms/unix/platform_unix.cpp
index a96daf7..dc957ec 100644
--- a/platforms/unix/platform_unix.cpp
+++ b/platforms/unix/platform_unix.cpp
@@ -159,18 +159,31 @@ void PlatformUnix::alterItem(CatItem* item) {
     QString exe = "";
     while(!file.atEnd()) {
 	QString line = file.readLine();
-	
-	if (line.startsWith("Name[" + locale, Qt::CaseInsensitive)) 
-	    name = line.split("=")[1].trimmed();
-	
-
-	else if (line.startsWith("Name=", Qt::CaseInsensitive)) 
-	    name = line.split("=")[1].trimmed();
-
-	else if (line.startsWith("Icon", Qt::CaseInsensitive))
-	    icon = line.split("=")[1].trimmed();
-	else if (line.startsWith("Exec", Qt::CaseInsensitive))
-	    exe = line.section("=", 1).trimmed();
+
+        if (name == "") {
+            if (line.startsWith("Name[" + locale, Qt::CaseInsensitive)) {
+                name = line.split("=")[1].trimmed();
+                continue;
+            }
+            else if (line.startsWith("Name=", Qt::CaseInsensitive)) {
+                name = line.split("=")[1].trimmed();
+                continue;
+            }
+        }
+
+        if (icon == "") {
+            if (line.startsWith("Icon", Qt::CaseInsensitive)) {
+                icon = line.split("=")[1].trimmed();
+                continue;
+            }
+        }
+
+        if (exe == "") {
+	    if (line.startsWith("Exec", Qt::CaseInsensitive)) {
+                exe = line.section("=", 1).trimmed();
+                continue;
+            }
+        }
     }
     if (name.size() >= item->shortName.size() - 8) {
 	item->shortName = name;
